core/migration.go

```diff
@@ -0,0 +1,174 @@
+
+// UpdateGroup is a type of modification that attempts to create or update a group to the given name and ID
+type UpdateGroup struct {
+	GroupName  string `toml:"name"`
+	NewGroupID int    `toml:"id"`
+}
+
+// LoadMigrations finds migration files in SysDir and UsrDir and attempts to load them
+func LoadMigrations() []Migration {
+	var allMigrations = make([]Migration, 0)
+
+	if sysFiles, err := ioutil.ReadDir(SysDir); err != nil {
+		log.Warnf("System directory for migrations at %s is unreadable, skipping\n", SysDir)
+	} else {
+		log.Debugf("Loading migrations from system directory %s...\n", SysDir)
+		for _, it := range sysFiles {
+			allMigrations = appendMigrationFrom(allMigrations, SysDir, it.Name())
+		}
+	}
+
+	if usrFiles, err := ioutil.ReadDir(UsrDir); err != nil {
+		log.Warnf("User directory for migrations at %s is unreadable, skipping\n", UsrDir)
+	} else {
+		log.Debugf("Loading migrations from user directory %s...\n", UsrDir)
+		for _, it := range usrFiles {
+			allMigrations = appendMigrationFrom(allMigrations, UsrDir, it.Name())
+		}
+	}
+
+	return allMigrations
+}
+
+func appendMigrationFrom(migrations []Migration, dir string, name string) []Migration {
+	if migration, err := parseMigration(dir, name); err != nil {
+		log.Warnf("    Failed to parse migration %s: %s\n", name, err)
+	} else {
+		migrations = append(migrations, migration)
+		log.Debugf("    Loaded migration %s\n", name)
+	}
+	return migrations
+}
+
+func parseMigration(dir string, name string) (Migration, error) {
+	var migration Migration
+
+	// Read the configuration into the program
+	var path = fmt.Sprintf("%s/%s", dir, name)
+	var cfg, err = readFile(path)
+	if err != nil {
+		return migration, err
+	}
+
+	// Save the configuration into the content structure
+	if err := toml.Unmarshal(cfg, &migration); err != nil {
+		return migration, err
+	}
+
+	// Validate the migration so we don't load any invalid data
+	if err := migration.Validate(); err != nil {
+		return migration, err
+	}
+
+	migration.Name = name
+	migration.Path = path
+
+	return migration, nil
+}
+
+// Validate checks if a migration contains at least one modification
+func (m Migration) Validate() error {
+	if len(m.UpdateUsers) == 0 && len(m.UpdateGroup) == 0 {
+		return fmt.Errorf("migrations must contain at least one modification")
+	}
+	return nil
+}
+
+// Run applies the modifications contained in a migration
+func (m Migration) Run(context *Context) {
+	log.Debugf("Running migration %s...\n", m.Name)
+	for _, task := range m.UpdateUsers {
+		m.updateUsers(context, task)
+	}
+	for _, task := range m.UpdateGroup {
+		m.updateGroup(context, task)
+	}
+}
+
+func (m Migration) updateUsers(context *Context, task UpdateUsers) {
+	var filtered = context.FilterUsers(task.UserFilters...)
+
+	for _, user := range filtered {
+		if ran, err := context.AddToGroup(user, task.GroupName); err != nil {
+			log.Warnf("    Failed to add group %s to user %s due to error: %s\n", task.GroupName, user.Name, err)
```

Comment:
Should use the tab character "\t" instead of spaces or a literal tab in print statements.